src/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -128,37 +226,89 @@ public Assembly LoadFromStream(Stream assembly, Stream assemblySymbols)
+                {
+                    iSymbolLength = (int) assemblySymbols.Length;
+                    arrSymbols = new byte[iSymbolLength];
+
+                    assemblySymbols.Read(arrSymbols, 0, iSymbolLength);
+                }
 
-            // Get the symbol stream in byte[] if provided
-            byte[] arrSymbols = null;
-            if (assemblySymbols != null)
+                RuntimeAssembly loadedAssembly = null;
+                unsafe
+                {
+                    fixed (byte* ptrAssembly = arrAssembly, ptrSymbols = arrSymbols)
+                    {
+                        LoadFromStream(m_pNativeAssemblyLoadContext, new IntPtr(ptrAssembly), iAssemblyStreamLength,
+                            new IntPtr(ptrSymbols), iSymbolLength, JitHelpers.GetObjectHandleOnStack(ref loadedAssembly));
+                    }
+                }
+                return loadedAssembly;
+#if FEATURE_COLLECTIBLE_ALC
+            }
+#endif
+        }
+
+        public void Unload()
+        {
+            if (!IsCollectible)
             {
-                iSymbolLength = (int)assemblySymbols.Length;
-                arrSymbols = new byte[iSymbolLength];
+                throw new InvalidOperationException(SR.GetResourceString("AssemblyLoadContext_Unload_CannotUnloadIfNotCollectible"));
+            }
 
-                assemblySymbols.Read(arrSymbols, 0, iSymbolLength);
+#if FEATURE_COLLECTIBLE_ALC
+            lock (unloadLock)
+            {
+                UnloadCollectible();
             }
+#endif
+        }
 
-            RuntimeAssembly loadedAssembly = null;
-            unsafe
+#if FEATURE_COLLECTIBLE_ALC
+        private void UnloadCollectible()
+        {
+            Debug.Assert(IsCollectible);
+            if (state == InternalState.Alive)
             {
-                fixed (byte* ptrAssembly = arrAssembly, ptrSymbols = arrSymbols)
-                {
-                    LoadFromStream(m_pNativeAssemblyLoadContext, new IntPtr(ptrAssembly), iAssemblyStreamLength, new IntPtr(ptrSymbols), iSymbolLength, JitHelpers.GetObjectHandleOnStack(ref loadedAssembly));
-                }
+                // Only if this ALC is collectible
+                var thisStrongHandle = GCHandle.Alloc(this, GCHandleType.Normal);
+                var thisStrongHandlePtr = GCHandle.ToIntPtr(thisStrongHandle);
+
+                // The underlying code will transform the original weak handle 
+                // created by InitializeLoadContext to a strong handle
+                PrepareForAssemblyLoadContextRelease(m_pNativeAssemblyLoadContext, thisStrongHandlePtr);
+            }
+            else
+            {
+                throw new InvalidOperationException(SR.GetResourceString("AssemblyLoadContext_Unload_AlreadyUnloaded"));
             }
+            state = InternalState.Unloading;
+        }
 
-            return loadedAssembly;
+        private void VerifyIsAlive()
+        {
+            if (state != InternalState.Alive)
```

Comment:
When the unload event handlers are executing, their JITing may trigger loads of more assemblies into the AssemblyLoadContext. But this block will prevent those assemblies from being loaded because the state is not Alive anymore. I think we need to allow loading of more stuff into the assembly load context even when the unload event handlers are executing.

We should have a test for this.